packages/transaction-controller/src/gas-flows/OptimismLayer1GasFeeFlow.ts

```diff
@@ -16,23 +19,163 @@ const OPTIMISM_STACK_CHAIN_IDS: Hex[] = [
   '0x420000000000000000000000000000000000000F';
 
+// opBNB uses a different oracle
+const OPBNB_GAS_PRICE_ORACLE_ADDRESS =
+  '0xc0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d3000f';
+
+type SupportedNetworksResponse = {
+  fullSupport: number[];
+  partialSupport: {
+    optimism: number[];
+  };
+};
+
+const GAS_SUPPORTED_NETWORKS_ENDPOINT =
+  'https://gas.api.cx.metamask.io/v1/supportedNetworks';
+
+let cachedOptimismChains: Set<number> | null = null;
+let isFetchingOptimismChains = false;
+let latestMessenger: TransactionControllerMessenger | null = null;
+
+/**
+ * Fetches the list of OP-stack chain IDs that the gas service currently supports.
+ *
+ * This is used by `ensureOptimismChainsFetch` and indirectly by
+ * `matchesTransaction` to decide whether a transaction's chain should be treated
+ * as Optimism-stack by consulting the backend, and to publish
+ * `OPTIMISM_CHAINS_UPDATED_EVENT` when the set changes.
+ *
+ * If the request fails, `null` is returned so callers can gracefully fall back
+ * to the local static list (`FALLBACK_OPTIMISM_STACK_CHAIN_IDS`).
+ *
+ * @returns A set of supported OP-stack chain IDs, or `null` if unavailable.
+ */
+async function fetchOptimismSupportedChains(): Promise<Set<number> | null> {
+  try {
+    const res: SupportedNetworksResponse = await handleFetch(
+      GAS_SUPPORTED_NETWORKS_ENDPOINT,
+    );
+    const list = res?.partialSupport?.optimism ?? [];
+    return new Set<number>(list);
+  } catch {
+    return null;
+  }
+}
+
+/**
+ * Determines whether two sets contain exactly the same elements.
+ *
+ * Used to detect changes to the cached set of OP-stack chain IDs. A `null`
+ * previous set is considered unequal to any non-null set so that the first
+ * successful fetch is treated as a change.
+ *
+ * @param first - The previous set of numeric chain IDs, or `null` if unset.
+ * @param second - The new set of numeric chain IDs to compare against.
+ * @returns `true` if both sets are non-null and have identical contents; otherwise `false`.
+ */
+function setsEqual(first: Set<number> | null, second: Set<number>): boolean {
+  if (!first || first.size !== second.size) {
+    return false;
+  }
+
+  for (const value of second) {
+    if (!first.has(value)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Ensures a single background fetch of supported OP-stack chain IDs is running.
+ *
+ * If there is no cached set and no fetch in progress, this starts a background
+ * request via `fetchOptimismSupportedChains()`. When the fetched set differs
+ * from the cached one, the cache is updated and `OPTIMISM_CHAINS_UPDATED_EVENT`
+ * is published using the latest provided messenger.
+ *
+ * Supplying a `messenger` stores it for subsequent event publication. Safe to
+ * call repeatedly; it is a no-op while a fetch is in flight or after the cache
+ * has been populated.
+ *
+ * @param messenger - Optional messenger used to publish update events.
+ */
+function ensureOptimismChainsFetch(
```

Comment:
Minor, ideally any utils are at the bottom of the file, or just private methods inside the class if not needed by anything else?